### PR TITLE
Initial Framework Philosophy document

### PR DESCRIPTION
**Pull request by vontell**

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+bot.openAndInteractWithChest(chest, (chest) => {
+    chest.place(...)
+});
+```
+
+In the cases of these context-aware operations (opening and closing a chest, mounting and unmounting a horse, etc..) we should still provide the context-unaware functions, that the user can use themselves). These are useful for sending new players quick code examples that they can understand.
+
+**ALSO DO**
+```javascript
+let openChest = bot.openChest(chest);
```
Comment (RG-nAmKcAz) [inline CONTRIBUTING.md:150]: we should should then how to use try/finally for any cases like this where the follow up API call is required to properly manage state

**Review by nAmKcAz — CHANGES_REQUESTED** [review]
Overall I like this a-lot. I have a few pieces of feedback in the examples themselves, but don't let that take away from the overall concept which I very much like / agree with.

I feel like this addresses some of my short term questions about the library and maybe some of @abeizer 's as well ?

I am still curious though on medium to long term use of the library and what Facts/Tests/etc we could add to this to make sure we stay roughly on target for the future.  Examples... If we plan to support block coding, what Facts/Tests can we add to ensure we keep that on track ?  If we plan to do multiple games or 3rd party integrations to games, what do we need to keep in mind to keep that on track ?

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,173 @@
+# Framework Philosophies
+
+This document outlines the goals, ideas, and roadmap for the rg-bot library (as well as our public-facing libraries in general). This is a living document, and we encourage players to participate and contribute to this document.
+
+## Core Facts
+
+The success of this library is dictated by various core facts that should be true. Our goal is to make sure that these facts are always true, and any change that causes one of these facts to not be true should be re-evaluated.
```
Comment (abeizer) [inline CONTRIBUTING.md:7]: Sounds a bit wonky because facts are implicitly true. Perhaps something along the lines of 

"The success of this library is dictated by its adherence to several core facts. Our goal is to make sure that these facts inform changes to the library. Any change that violates one or more of these facts must be re-evaluated."

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+
+Note that some of these facts require testing the library against real users. We don't need to evaluate these specific items every time a change occurs in the library, but we should make that evaluation often.
```
Comment (abeizer) [inline CONTRIBUTING.md:9]: I'm not sure what you're getting at here from a Contributor perspective. Is this intended to state that when validation against our facts requires access to a production environment that we (the RG team) will be responsible for that validation? It sounds more like you're stating that not every change needs to be evaluated against our core facts, but that defeats the purpose of having them so I'm a bit confused on this one.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+
+**Measurement Approach:** Survey / Interview
+
+**Desired sample size per test**: 10-15 users
+
+**Most recent test results:** We should include the **average** of all the examples, and a table of the specific templates and examples (i.e. each example bot).
+
+Example Prompt for the Interviewee
+```
+Please describe what this bot does. Try to use a full sentence, and cover (at a high level) the entire algorithm, not just one piece of it.
```
Comment (abeizer) [inline CONTRIBUTING.md:73]:
```suggestion
Please describe what this bot does. Try to use a full sentence, and cover (at a high level) the entire algorithm rather than just one piece of it.
```

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+```
+
+### Fact 2: 95% of users with great than 1 year of JavaScript experience should be able to articulate what the provided examples / templates do
```
Comment (abeizer) [inline CONTRIBUTING.md:76]:
```suggestion
### Fact 2: 95% of users with more than 1 year of JavaScript experience should be able to articulate what the provided examples / templates do
```

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+Making this goal a reality ensures that players who know JavaScript well can jump right in and begin understanding the bots on the Regression Games platform. This is measured by conducting user-interviews and surveys with potential players. The user should be able to _roughly describe_ the algorithm in the example (we define this as getting most of the algorithm).
```
Comment (abeizer) [inline CONTRIBUTING.md:77]:
```suggestion
This fact ensures that players with an intermediate level of coding experience can engage with and understand  bots on the Regression Games platform. This goal is measured by conducting interviews and surveys with players and potential players. The player should be able to _roughly describe_ the algorithm in a given example -we define this as understanding most of the algorithm.
```

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+
+This test is the same as **Fact 1**, but has a higher percentage of players since the players have more experience with JavaScript.
```
Comment (abeizer) [inline CONTRIBUTING.md:79]:
```suggestion
This test is the same as **Fact 1** but requires a higher success rate because the targeted players are expected to be more experienced with JavaScript.
```

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+
+**Measurement Approach:** Survey / Interview
+
+**Desired sample size per test**: 10-15 users
+
+**Most recent test results:** We should include the **average** of all the examples, and a table of the specific templates and examples (i.e. each example bot).
```
Comment (abeizer) [inline CONTRIBUTING.md:85]: Same comment as fact 1

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+```javascript
+const animalsToHunt = ["chicken", "pig", "cow", "sheep", "player"];
+
+async function huntAnimals() {
+    let nearbyAnimals = bot.findEntities({entityNames: animalsToHunt, maxDistance: 100});
+    console.log(nearbyAnimals)
+    if (nearbyAnimals.length == 0) {
+        bot.chat("Could not find animals nearby... going to wander and try again")
+        await bot.wander();
+        return await huntAnimals()
```
[the remark this answers is not in the document]
Comment (vontell) [reply CONTRIBUTING.md]: Resolved - returned if no animals were found (with a helpful message), and added a retry limit to the attacking procedure.

File: CONTRIBUTING.md (hunk 1 above)
Comment (vontell) [reply CONTRIBUTING.md]: Added!

File: CONTRIBUTING.md (hunk 2 above)
Comment (vontell) [reply CONTRIBUTING.md]: Fixed!

File: CONTRIBUTING.md (hunk 3 above)
Comment (vontell) [reply CONTRIBUTING.md:9]: By this suggestion, I meant that it is probably not feasible to run these surveys with every change we make, so I wanted to give some leeway so that we can use our best judgment, and batch multiple changes into one survey. Thoughts?

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+This test is the same as **Fact 1**, but has a higher percentage of players since the players have more experience with JavaScript.
+
+**Measurement Approach:** Survey / Interview
+
+**Desired sample size per test**: 10-15 users
+
+**Most recent test results:** We should include the **average** of all the examples, and a table of the specific templates and examples (i.e. each example bot).
+
+
+### Fact 3: A user that encounters an error within our library that they must recover from should immediately have one next step available to them
```
Comment (abeizer) [inline CONTRIBUTING.md:88]: Descriptive and accurate, but wording doesn't flow very well. Suggest something like...

"When an error results from correct usage of our library, the library must offer the user at least one option for recovery from the error"

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+Focusing on this goal ensures that we increase the chance that a player never gets stuck on a library-related issue. Players should spend their time focusing on their strategies, rather than the intricacies of code. When a fatal error occurs that requires the user to fix their code, we should make every attempt to point them in the right direction. Examples of this include:
```
Comment (abeizer) [inline CONTRIBUTING.md:89]:
```suggestion
This fact ensures that a player never gets stuck due to a library-related issue. Players should be able to focus on their strategies rather than the intricacies of the library. When a fatal error occurs which requires handling in the user's code, we must attempt to point them in the right direction towards resolution. Examples of this include:
```

File: CONTRIBUTING.md [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+## Core Facts
+
+The success of this library is dictated by various core facts that should be true. Our goal is to make sure that these facts are always true, and any change that causes one of these facts to not be true should be re-evaluated.
+
+Note that some of these facts require testing the library against real users. We don't need to evaluate these specific items every time a change occurs in the library, but we should make that evaluation often.
+
+#### Fact 1: 80% of users with more than 6 months and less than 1 year of JavaScript experience should be able to articulate what a provided example / template does.
+Making this goal a reality ensures that most new players who may not be coding experts can jump right in and begin understanding the bots on the Regression Games platform. This is measured by conducting user-interviews and surveys with potential players. The user should be able to _roughly describe_ the algorithm in the example (we define this as getting most of the algorithm). Here is an example of a template:
+
+```javascript
```
Comment (abeizer) [inline CONTRIBUTING.md:14]: Title this as "prompt", "template", etc. to visually separate it from the answers/responses section

File: CONTRIBUTING.md [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+**Measurement Approach:** Survey / Interview
+
+**Desired sample size per test**: 10-15 users
+
+**Most recent test results:** We should include the **average** of all the examples, and a table of the specific templates and examples (i.e. each example bot).
+
+
+### Fact 3: A user that encounters an error within our library that they must recover from should immediately have one next step available to them
+Focusing on this goal ensures that we increase the chance that a player never gets stuck on a library-related issue. Players should spend their time focusing on their strategies, rather than the intricacies of code. When a fatal error occurs that requires the user to fix their code, we should make every attempt to point them in the right direction. Examples of this include:
+* Logging errors that make direct suggestions to fix (e.g. "Invalid item name. Make sure that the item name is a valid id (e.g 'spruce_log') instead of the displayed name (e.g. 'Spruce Log'). The list of valid items can be found at https://...")
```
Comment (abeizer) [inline CONTRIBUTING.md:90]:
```suggestion
* Logging errors with a direct suggestion on how to fix the root cause of the error (e.g. "Invalid item name. Make sure that the item name is a valid id (e.g 'spruce_log') instead of a displayed name (e.g. 'Spruce Log'). The list of valid items can be found at https://...")
```

File: CONTRIBUTING.md [hunk 14]
```diff
@@ -0,0 +1,173 @@
...
+* Logging errors that include links to documentaton for next steps (e.g. "... To see approaches to fix this, visit https://...")
```
Comment (abeizer) [inline CONTRIBUTING.md:91]:
```suggestion
* Logging errors that include links to documentation for next steps (e.g. "... To see approaches to fix this, visit https://...")
```

File: CONTRIBUTING.md [hunk 15]
```diff
@@ -0,0 +1,173 @@
...
+* Explaining it more simple terms **why** it failed
```
Comment (abeizer) [inline CONTRIBUTING.md:92]: * `it more` -> `in more`
* `it failed` -> `the action failed`
* Can you provide an example here?

File: CONTRIBUTING.md [hunk 16]
```diff
@@ -0,0 +1,173 @@
...
+
+Good Examples of handling this:
```
Comment (abeizer) [inline CONTRIBUTING.md:94]:
```suggestion
Acceptable examples of providing recovery options to the user:
```

File: CONTRIBUTING.md [hunk 17]
```diff
@@ -0,0 +1,173 @@
...
+* [Explicitly stating that the error is fine and expected](https://github.com/Regression-Games/RegressionBot/blob/e2bddcc37950a9f2ee5718ff5aa4dda549c5f008/lib/RGBot.js#L1088)
+* [Saying why it failed in simple turns](https://github.com/Regression-Games/RegressionBot/blob/e2bddcc37950a9f2ee5718ff5aa4dda549c5f008/lib/RGBot.js#L1102)
```
Comment (abeizer) [inline CONTRIBUTING.md:96]:
```suggestion
* [Stating why an action failed in simple terms](https://github.com/Regression-Games/RegressionBot/blob/e2bddcc37950a9f2ee5718ff5aa4dda549c5f008/lib/RGBot.js#L1102)
```

File: CONTRIBUTING.md [hunk 18]
```diff
@@ -0,0 +1,173 @@
...
+
+
+### Fact 3: A user that encounters an error within our library that they must recover from should immediately have one next step available to them
+Focusing on this goal ensures that we increase the chance that a player never gets stuck on a library-related issue. Players should spend their time focusing on their strategies, rather than the intricacies of code. When a fatal error occurs that requires the user to fix their code, we should make every attempt to point them in the right direction. Examples of this include:
+* Logging errors that make direct suggestions to fix (e.g. "Invalid item name. Make sure that the item name is a valid id (e.g 'spruce_log') instead of the displayed name (e.g. 'Spruce Log'). The list of valid items can be found at https://...")
+* Logging errors that include links to documentaton for next steps (e.g. "... To see approaches to fix this, visit https://...")
+* Explaining it more simple terms **why** it failed
+
+Good Examples of handling this:
+* [Explicitly stating that the error is fine and expected](https://github.com/Regression-Games/RegressionBot/blob/e2bddcc37950a9f2ee5718ff5aa4dda549c5f008/lib/RGBot.js#L1088)
```
Comment (abeizer) [inline CONTRIBUTING.md:95]:
```suggestion
* [Explicitly stating that an error is expected](https://github.com/Regression-Games/RegressionBot/blob/e2bddcc37950a9f2ee5718ff5aa4dda549c5f008/lib/RGBot.js#L1088)
```

File: CONTRIBUTING.md [hunk 19]
```diff
@@ -0,0 +1,173 @@
...
+* [Saying why it failed in simple turns](https://github.com/Regression-Games/RegressionBot/blob/e2bddcc37950a9f2ee5718ff5aa4dda549c5f008/lib/RGBot.js#L1102)
+
+**Measurement Approach:** Internal Audit
+
+**Most recent test results:** 
+
+(Dec 27, 2022) Out of **17** explicit errors, **9** had a next steps or more info provided = **53%** success rate.
```
Comment (abeizer) [inline CONTRIBUTING.md:102]:
```suggestion
(Dec 27, 2022) Out of **17** explicit errors, **9** provided next steps or more info = **53%** success rate.
```

**Review by abeizer — CHANGES_REQUESTED** [review]

File: CONTRIBUTING.md (hunk 3 above)
Comment (abeizer) [reply CONTRIBUTING.md:9]: That's reasonable - gathering feedback from users will definitely take time. With this in mind, I'd suggest rewording to something like...

"Some of the following core facts require gathering feedback from real users, which is not feasible as a requirement for accepting every change to the library as it is proposed. Therefore, it is the responsibility of the Regression Games team to evaluate compliance with these facts at its discretion on a regular basis."

File: CONTRIBUTING.md [hunk 20]
```diff
@@ -0,0 +1,183 @@
...
+
+## Link to Survey for Fact Checking
+
+https://d7y6yysps34.typeform.com/to/feG9z7kp
+
+## Design Patterns
+
+### Awaits, Promises, and Callbacks
+
+Most of our functions and design patterns should use Promises + `await` when possible. For example:
```
Comment (abeizer) [inline CONTRIBUTING.md:122]:
```suggestion
Most of our functions and design patterns should use an `async/await` pattern for Promises when possible. For example:
```

File: CONTRIBUTING.md [hunk 21]
```diff
@@ -0,0 +1,183 @@
...
+        if (attackCount > 100) {
+            bot.chat("Could not kill animal, stopping. Try hunting a new animal.")
+            return;
+        }
+        await bot.attackEntity(animalToAttack)
+    }
+    bot.chat("Finished attacking the " + animalToAttack.type, ", moving on the next victim")
+    let itemsOnGround = await bot.findAndCollectItemsOnGround()
+    bot.chat(`Picked up ${itemsOnGround.length} items off the ground`)
+    return await huntAnimals()
```
Comment (RG-nAmKcAz) [inline CONTRIBUTING.md:37]: I would suggest using the code sample , or very close to it, from the replit example.

https://linear.app/regression-games/issue/REG-505#comment-cdf59209

File: CONTRIBUTING.md [hunk 22]
```diff
@@ -0,0 +1,183 @@
...
+* [When dropping an item fails](https://github.com/Regression-Games/RegressionBot/blob/e2bddcc37950a9f2ee5718ff5aa4dda549c5f008/lib/RGBot.js#L1417)
+* [Error when crafting an item](https://github.com/Regression-Games/RegressionBot/blob/e2bddcc37950a9f2ee5718ff5aa4dda549c5f008/lib/RGBot.js#L1498)
+
+## Link to Survey for Fact Checking
+
+https://d7y6yysps34.typeform.com/to/feG9z7kp
+
+## Design Patterns
+
+### Awaits, Promises, and Callbacks
```
Comment (abeizer) [inline CONTRIBUTING.md:120]:
```suggestion
### Promises and Callbacks
```

File: CONTRIBUTING.md [hunk 23]
```diff
@@ -0,0 +1,183 @@
...
+In the case where a user-provided block of code **needs** to be passed to a library function, this is fine. This should really only happen when the operation requires some sort of context that needs managing (such as opening a chest, e.g. there is a before and after operation). For example:
+
+**DO**
+```javascript
+bot.openAndInteractWithChest(chest, (chest) => {
+    chest.place(...)
+});
+```
+
+In the cases of these context-aware operations (opening and closing a chest, mounting and unmounting a horse, etc..) we should still provide the context-unaware functions, that the user can use themselves). These are useful for sending new players quick code examples that they can understand.
```
Comment (abeizer) [inline CONTRIBUTING.md:150]:
```suggestion
In the cases of these context-aware operations (opening and closing a chest, mounting and unmounting a horse, etc..) we should still provide the context-unaware functions, which the user can use themselves). These are useful for sending new players quick code examples that they can understand.
```

**Comment by abeizer** [comment]
> Overall I like this a-lot. I have a few pieces of feedback in the examples themselves, but don't let that take away from the overall concept which I very much like / agree with.
> 
> I feel like this addresses some of my short term questions about the library and maybe some of @abeizer 's as well ?
> 
> I am still curious though on medium to long term use of the library and what Facts/Tests/etc we could add to this to make sure we stay roughly on target for the future. Examples... If we plan to support block coding, what Facts/Tests can we add to ensure we keep that on track ? If we plan to do multiple games or 3rd party integrations to games, what do we need to keep in mind to keep that on track ?

@vontell Agree with Zack that I like where this is going and it's great that we have a place to start outlining our expectations for this library, but there are bits we'd benefit from discussing + adding. 

For example, we've talked internally about the importance of having both 'out of the box' methods as well as making behaviors configurable for more advanced users. This concept and the practice of using `options` to achieve it is something that should definitely be outlined somewhere in this document.

We also already know from internal discussions that our requirements re: user accessibility _will_ shift as we explore different methods of moving from code to no-code. I saw your disclaimer about this being a living document but I'm not sure whether the expectation is for the facts themselves to change over time or if the facts are expected to stay set in stone while our methods and patterns for satisfying them is the 'living' aspect.

**Comment by abeizer** [comment]
Another thing we should add: practices for deprecation warnings, following the pattern that Zack put in place

**Comment by abeizer** [comment]
Can we close this so it stops showing in my pull requests summary?

**Comment by vontell** [comment]
@abeizer sure thing XD
